File: Biblioteca-main/Backend/models/Reserva.js
const mongoose = require('mongoose');

const reservaSchema = new mongoose.Schema({
  user_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  libro_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Libro',
    required: [true, 'El ID del libro es obligatorio']
  },
  tipo: {
    type: String,
    enum: ['prestamo', 'sala', 'domicilio'],
    default: 'prestamo'
  },
  desde: {
    type: Date,
    required: [true, 'La fecha de inicio es obligatoria']
  },
  hasta: {
    type: Date,
    required: [true, 'La fecha de fin es obligatoria']
  },
  estado: {
    type: String,
    enum: ['activa', 'finalizada', 'cancelada'],
    default: 'activa'
  }
}, {
  timestamps: true
});

reservaSchema.pre('save', function(next) {
  if (this.hasta <= this.desde) {
    next(new Error('La fecha de fin debe ser posterior a la fecha de inicio'));
  }
  next();
});

module.exports = mongoose.model('Reserva', reservaSchema);

File: Biblioteca-main/Movil/biblioteca_flutter/lib/models/reserva.dart
import 'libro.dart';
import 'user.dart';

class Reserva {
  final String id;
  final String userId;
  final String libroId;
  final String tipo;
  final DateTime desde;
  final DateTime hasta;
  final String estado;
  final Libro? libro;
  final User? user;

  Reserva({
    required this.id,
    required this.userId,
    required this.libroId,
    required this.tipo,
    required this.desde,
    required this.hasta,
    required this.estado,
    this.libro,
    this.user,
  });

  factory Reserva.fromJson(Map<String, dynamic> json) {
    return Reserva(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      userId: json['user_id']?.toString() ?? '',
      libroId: json['libro_id']?.toString() ?? '',
      tipo: json['tipo'] ?? 'prestamo',
      desde: json['desde'] != null
          ? DateTime.parse(json['desde'])
          : DateTime.now(),
      hasta: json['hasta'] != null
          ? DateTime.parse(json['hasta'])
          : DateTime.now(),
      estado: json['estado'] ?? 'activa',
      libro: json['libro_id'] is Map
          ? Libro.fromJson(json['libro_id'])
          : (json['libro'] != null ? Libro.fromJson(json['libro']) : null),
      user: json['user_id'] is Map ? User.fromJson(json['user_id']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'libro_id': libroId,
      'tipo': tipo,
      'desde': desde.toIso8601String().split('T')[0],
      'hasta': hasta.toIso8601String().split('T')[0],
    };
  }
}
